From be2a509686807ea9e10e911b582828b642b5f274 Mon Sep 17 00:00:00 2001
From: Zhang Sanshan <b51434@freescale.com>
Date: Wed, 17 Jun 2015 20:55:14 +0800
Subject: [PATCH] MA-6803 enable 7D touch in lp5.1

The coordinate of top left and buttom right exchange for 7d sdb hardware error.
It correct the X\Y value which is reported to user space according dts.

Signed-off-by: Zhang Sanshan <b51434@freescale.com>
---
 arch/arm/boot/dts/imx7d-sdb-touch.dts | 1 +
 drivers/input/touchscreen/ads7846.c   | 7 ++++++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boot/dts/imx7d-sdb-touch.dts b/arch/arm/boot/dts/imx7d-sdb-touch.dts
index b53f0be..dc5b276 100644
--- a/arch/arm/boot/dts/imx7d-sdb-touch.dts
+++ b/arch/arm/boot/dts/imx7d-sdb-touch.dts
@@ -26,6 +26,7 @@
 		ti,y-max = <4800>;
 		ti,pressure-max = <255>;
 		ti,x-plat-ohms = <40>;
+		ti,yx_swap_en;
 		linux,wakeup;
 	};
 };
diff --git a/drivers/input/touchscreen/ads7846.c b/drivers/input/touchscreen/ads7846.c
index 584113a..a71c285 100644
--- a/drivers/input/touchscreen/ads7846.c
+++ b/drivers/input/touchscreen/ads7846.c
@@ -199,6 +199,7 @@ struct ads7846 {
 #define	REF_ON	(READ_12BIT_DFR(x, 1, 1))
 #define	REF_OFF	(READ_12BIT_DFR(y, 0, 0))
 
+static bool  rotation_180;
 /* Must be called with ts->lock held */
 static void ads7846_stop(struct ads7846 *ts)
 {
@@ -822,6 +823,10 @@ static void ads7846_report_state(struct ads7846 *ts)
 	 * timer by reading the pen signal state (it's a GPIO _and_ IRQ).
 	 */
 	if (Rt) {
+		if (rotation_180) {
+			x = MAX_12BIT - x;
+			y = MAX_12BIT - y;
+		}
 		struct input_dev *input = ts->input;
 
 		if (ts->swap_xy)
@@ -1235,7 +1240,7 @@ static const struct ads7846_platform_data *ads7846_probe_dt(struct device *dev)
 
 	of_property_read_u32(node, "ti,pendown-gpio-debounce",
 			     &pdata->gpio_pendown_debounce);
-
+	rotation_180 = of_property_read_bool(node, "ti,yx_swap_en");
 	pdata->wakeup = of_property_read_bool(node, "linux,wakeup");
 
 	pdata->gpio_pendown = of_get_named_gpio(dev->of_node, "pendown-gpio", 0);
-- 
1.8.0

